From 23289c1be0cb1d5c371e5cdbff282f6c3185223d Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Tor=20Arne=20Vestb=C3=B8?= <tor.arne.vestbo@qt.io>
Date: Thu, 30 Jul 2026 19:53:54 +0200
Subject: [PATCH] macOS: Make the isMouseEvent helper available to the whole
 plugin

It was a file static in qnswindow.mm, but we need the same check in
other places as well. We now express it in terms of the existing
cocoaEvent2QtMouseEvent, which as a side effect also covers the
NSEventTypeOtherMouse* events that the file static left out, so frame
strut handling and popup redirection now treat the middle mouse button
like the left and right ones.

Pick-to: 6.8
Change-Id: I244d962fcc52b0bbb43ec66a2860ee55a752d2ce
Reviewed-by: Doris Verria <doris.verria@qt.io>
(cherry picked from commit fd22838f0efb997da3462c225538508e603e201f)
Reviewed-by: Qt Cherry-pick Bot <cherrypick_bot@qt-project.org>
(cherry picked from commit 5d6ed915ec7630951fd4c480ec69ab5b94b817b7)
---
 src/plugins/platforms/cocoa/qcocoahelpers.h  |  1 +
 src/plugins/platforms/cocoa/qcocoahelpers.mm |  5 +++++
 src/plugins/platforms/cocoa/qnswindow.mm     | 20 ++------------------
 3 files changed, 8 insertions(+), 18 deletions(-)

diff --git a/src/plugins/platforms/cocoa/qcocoahelpers.h b/src/plugins/platforms/cocoa/qcocoahelpers.h
index 0e67629177cd..bc89257a1aba 100644
--- a/src/plugins/platforms/cocoa/qcocoahelpers.h
+++ b/src/plugins/platforms/cocoa/qcocoahelpers.h
@@ -73,6 +73,7 @@ Qt::MouseButton cocoaButton2QtButton(NSInteger buttonNum);
 Qt::MouseButton cocoaButton2QtButton(NSEvent *event);
 
 QEvent::Type cocoaEvent2QtMouseEvent(NSEvent *event);
+bool qt_mac_isMouseEvent(NSEvent *event);
 
 Qt::MouseButtons cocoaMouseButtons2QtMouseButtons(NSInteger pressedMouseButtons);
 Qt::MouseButtons currentlyPressedMouseButtons();
diff --git a/src/plugins/platforms/cocoa/qcocoahelpers.mm b/src/plugins/platforms/cocoa/qcocoahelpers.mm
index e86386aad028..f05fb333d715 100644
--- a/src/plugins/platforms/cocoa/qcocoahelpers.mm
+++ b/src/plugins/platforms/cocoa/qcocoahelpers.mm
@@ -309,6 +309,11 @@ QRectF qt_mac_flip(const QRectF &rect, const QRectF &reference)
     return QEvent::None;
 }
 
+bool qt_mac_isMouseEvent(NSEvent *event)
+{
+    return cocoaEvent2QtMouseEvent(event) != QEvent::None;
+}
+
 /*!
   \fn Qt::MouseButtons cocoaMouseButtons2QtMouseButtons(NSInteger pressedMouseButtons)
 
diff --git a/src/plugins/platforms/cocoa/qnswindow.mm b/src/plugins/platforms/cocoa/qnswindow.mm
index d678b953df7f..e65641944519 100644
--- a/src/plugins/platforms/cocoa/qnswindow.mm
+++ b/src/plugins/platforms/cocoa/qnswindow.mm
@@ -18,22 +18,6 @@
 
 Q_STATIC_LOGGING_CATEGORY(lcQpaEvents, "qt.qpa.events");
 
-static bool isMouseEvent(NSEvent *ev)
-{
-    switch ([ev type]) {
-    case NSEventTypeLeftMouseDown:
-    case NSEventTypeLeftMouseUp:
-    case NSEventTypeRightMouseDown:
-    case NSEventTypeRightMouseUp:
-    case NSEventTypeMouseMoved:
-    case NSEventTypeLeftMouseDragged:
-    case NSEventTypeRightMouseDragged:
-        return true;
-    default:
-        return false;
-    }
-}
-
 @implementation NSWindow (FullScreenProperty)
 
 + (void)load
@@ -342,7 +326,7 @@ - (void)sendEvent:(NSEvent*)theEvent
     }
 
     const bool mouseEventInFrameStrut = [theEvent, self]{
-        if (isMouseEvent(theEvent)) {
+        if (qt_mac_isMouseEvent(theEvent)) {
             const NSPoint loc = theEvent.locationInWindow;
             const NSRect windowFrame = [self convertRectFromScreen:self.frame];
             const NSRect contentFrame = self.contentView.frame;
@@ -366,7 +350,7 @@ - (void)sendEvent:(NSEvent*)theEvent
     // not Qt). However, an active popup is expected to grab any mouse event within the
     // application, so we need to handle those explicitly and trust Qt's isWindowBlocked
     // implementation to eat events that shouldn't be delivered anyway.
-    if (isMouseEvent(theEvent) && QGuiApplicationPrivate::instance()->activePopupWindow()
+    if (qt_mac_isMouseEvent(theEvent) && QGuiApplicationPrivate::instance()->activePopupWindow()
         && QGuiApplicationPrivate::instance()->isWindowBlocked(m_platformWindow->window(), nullptr)) {
         qCDebug(lcQpaWindow) << "Mouse event over modally blocked window" << m_platformWindow->window()
                              << "while popup is open - redirecting";
